File: win/directx/sdk-9.0/Samples/VB.Net/DirectPlay/DXMessenger/Client/wfMsg.vb
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms


Namespace DXMessengerClient
    _
    '/ <summary>
    '/ Summary description for wfMsg.
    '/ </summary>
    Public Class wfMsg
        Inherits System.Windows.Forms.Form

        Delegate Sub AddChatMsgCallback(ByVal msg As String, ByVal meTalking As Boolean, ByVal noTalking As Boolean)

        Delegate Sub ShowWindowCallback()
        Private txtMsg As System.Windows.Forms.TextBox
        Private WithEvents btnSend As System.Windows.Forms.Button
        '/ <summary>
        '/ Required designer variable.
        '/ </summary>
        Private components As System.ComponentModel.Container = Nothing
        Private WithEvents txtChat As System.Windows.Forms.TextBox

        Private msUser As String = Nothing
        Private Shadows parent As wfClient = Nothing


        Public ReadOnly Property UserName() As String
            Get
                Return msUser
            End Get
        End Property

        Public Sub New(ByVal username As String, ByVal obj As wfClient)
            '
            ' Required for Windows Form Designer support
            '
            InitializeComponent()

            msUser = username
            parent = obj
            Me.Text = "Message - " + msUser
        End Sub 'New


        '/ <summary>
        '/ Clean up any resources being used.
        '/ </summary>
        Protected Shadows Sub Dispose()
            If disposing Then
                If Not (components Is Nothing) Then
                    components.Dispose()
                End If
            End If
            MyBase.Dispose(disposing)
        End Sub 'Dispose


        '/ <summary>
        '/ Required method for Designer support - do not modify
        '/ the contents of this method with the code editor.
        '/ </summary>
        Private Sub InitializeComponent()
            Dim resources As New System.Resources.ResourceManager(GetType(wfMsg))
            Me.txtMsg = New System.Windows.Forms.TextBox()
            Me.txtChat = New System.Windows.Forms.TextBox()
            Me.btnSend = New System.Windows.Forms.Button()
            Me.SuspendLayout()
            ' 
            ' txtMsg
            ' 
            Me.txtMsg.Anchor = System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Bottom Or System.Windows.Forms.AnchorStyles.Left Or System.Windows.Forms.AnchorStyles.Right
            Me.txtMsg.Location = New System.Drawing.Point(6, 3)
            Me.txtMsg.Multiline = True
            Me.txtMsg.Name = "txtMsg"
            Me.txtMsg.ReadOnly = True
            Me.txtMsg.ScrollBars = System.Windows.Forms.ScrollBars.Both
            Me.txtMsg.Size = New System.Drawing.Size(454, 292)
            Me.txtMsg.TabIndex = 100
            Me.txtMsg.TabStop = False
            Me.txtMsg.Text = ""
            ' 
            ' txtChat
            ' 
            Me.txtChat.Anchor = System.Windows.Forms.AnchorStyles.Bottom Or System.Windows.Forms.AnchorStyles.Left Or System.Windows.Forms.AnchorStyles.Right
            Me.txtChat.Location = New System.Drawing.Point(4, 300)
            Me.txtChat.Multiline = True
            Me.txtChat.Name = "txtChat"
            Me.txtChat.ScrollBars = System.Windows.Forms.ScrollBars.Both
            Me.txtChat.Size = New System.Drawing.Size(377, 41)
            Me.txtChat.TabIndex = 0
            Me.txtChat.Text = ""
            ' 
            ' btnSend
            ' 
            Me.btnSend.Anchor = System.Windows.Forms.AnchorStyles.Bottom Or System.Windows.Forms.AnchorStyles.Right
            Me.btnSend.Location = New System.Drawing.Point(385, 302)
            Me.btnSend.Name = "btnSend"
            Me.btnSend.Size = New System.Drawing.Size(75, 37)
            Me.btnSend.TabIndex = 1
            Me.btnSend.Text = "Send"
            ' 
            ' wfMsg
            ' 
            Me.AcceptButton = Me.btnSend
            Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
            Me.ClientSize = New System.Drawing.Size(461, 341)
            Me.Controls.AddRange(New System.Windows.Forms.Control() {Me.btnSend, Me.txtChat, Me.txtMsg})
            Me.Icon = CType(resources.GetObject("$this.Icon"), System.Drawing.Icon)
            Me.Name = "wfMsg"
            Me.Text = "MessageTemplate"
            Me.ResumeLayout(False)
        End Sub 'InitializeComponent

        '
        Public Sub ShowWindow() '
            ' Show the window, and add the text
            Me.Show()
            Me.Select()
            Me.BringToFront()
        End Sub 'ShowWindow

        Private Sub wfMsg_Enter(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Enter
            ' We have the focus, make the chat textbox the default option
            Me.txtChat.Select()
        End Sub 'wfMsg_Enter


        Public Sub AddChatMessage(ByVal msg As String, ByVal meTalking As Boolean, ByVal noTalking As Boolean)
            If Not noTalking Then
                If meTalking Then
                    msg = "<" + parent.gUsername + "> " + msg
                Else
                    msg = "<" + msUser + "> " + msg
                End If
            End If
            ' Now limit the text to 32k
            If txtMsg.Text.Length > txtMsg.MaxLength * 0.95 Then
                txtMsg.Text = txtMsg.Text.Remove(0, CInt(txtMsg.MaxLength / 2))
            End If
            ' Update the message window
            txtMsg.Text += msg + ControlChars.Cr + ControlChars.Lf
            txtMsg.SelectionStart = txtMsg.Text.Length
            txtMsg.ScrollToCaret()
        End Sub 'AddChatMessage

        Private Sub btnSend_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnSend.Click
            ' We're talking
            If Not (txtChat.Text Is Nothing) And txtChat.Text <> "" Then
                If parent.gConnected Then
                    parent.SendChatMessage(msUser, txtChat.Text)
                    Dim unavailableMsg As Object() = {txtChat.Text, True, False}

                    Me.BeginInvoke(New AddChatMsgCallback(AddressOf Me.AddChatMessage), unavailableMsg)
                    txtChat.Text = Nothing
                Else
                    Dim unavailableMsg As Object() = {"**** - You are not connected to a server, you cannot send messages.", True, True}
                    Me.BeginInvoke(New AddChatMsgCallback(AddressOf Me.AddChatMessage), unavailableMsg)
                    txtChat.Text = Nothing
                End If
            End If
        End Sub 'btnSend_Click


        Private Sub txtChat_TextChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles txtChat.TextChanged
            ' Only allow them to send if there's data to be sent.
            btnSend.Enabled = Not (txtChat.Text Is Nothing) And txtChat.Text <> ""
        End Sub 'txtChat_TextChanged
    End Class 'wfMsg
End Namespace 'DXMessengerClient

File: win/directx/sdk-9.0/Samples/C#/DirectPlay/DXMessenger/Client/wfMsg.cs
//----------------------------------------------------------------------------
// File: wfMsg.cs
//
// Copyright (c) Microsoft Corp. All rights reserved.
//-----------------------------------------------------------------------------
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;

namespace DXMessengerClient
{
	/// <summary>
	/// Summary description for wfMsg.
	/// </summary>
	public class wfMsg : System.Windows.Forms.Form
	{
		public delegate void AddChatMsgCallback(string msg, bool meTalking, bool noTalking);
		public delegate void ShowWindowCallback();
		private System.Windows.Forms.TextBox txtMsg;
		private System.Windows.Forms.Button btnSend;
		/// <summary>
		/// Required designer variable.
		/// </summary>
		private System.ComponentModel.Container components = null;
		private System.Windows.Forms.TextBox txtChat;

		private string msUser = null;
		wfClient parent = null;

		public string UserName { get { return msUser; } }
		public wfMsg(string username, wfClient obj)
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();

			msUser = username;
			parent = obj;
			this.Text = "Message - "  + msUser;

		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose(bool disposing)
		{
			if (disposing)
			{
				if (components != null)
				{
					components.Dispose();
				}
			}
			base.Dispose(disposing);
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
			System.Resources.ResourceManager resources = new System.Resources.ResourceManager(typeof(wfMsg));
			this.txtMsg = new System.Windows.Forms.TextBox();
			this.txtChat = new System.Windows.Forms.TextBox();
			this.btnSend = new System.Windows.Forms.Button();
			this.SuspendLayout();
			// 
			// txtMsg
			// 
			this.txtMsg.Anchor = (((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Bottom) 
				| System.Windows.Forms.AnchorStyles.Left) 
				| System.Windows.Forms.AnchorStyles.Right);
			this.txtMsg.Location = new System.Drawing.Point(6, 3);
			this.txtMsg.Multiline = true;
			this.txtMsg.Name = "txtMsg";
			this.txtMsg.ReadOnly = true;
			this.txtMsg.ScrollBars = System.Windows.Forms.ScrollBars.Both;
			this.txtMsg.Size = new System.Drawing.Size(454, 292);
			this.txtMsg.TabIndex = 100;
			this.txtMsg.TabStop = false;
			this.txtMsg.Text = "";
			// 
			// txtChat
			// 
			this.txtChat.Anchor = ((System.Windows.Forms.AnchorStyles.Bottom | System.Windows.Forms.AnchorStyles.Left) 
				| System.Windows.Forms.AnchorStyles.Right);
			this.txtChat.Location = new System.Drawing.Point(4, 300);
			this.txtChat.Multiline = true;
			this.txtChat.Name = "txtChat";
			this.txtChat.ScrollBars = System.Windows.Forms.ScrollBars.Both;
			this.txtChat.Size = new System.Drawing.Size(377, 41);
			this.txtChat.TabIndex = 0;
			this.txtChat.Text = "";
			this.txtChat.TextChanged += new System.EventHandler(this.txtChat_TextChanged);
			// 
			// btnSend
			// 
			this.btnSend.Anchor = (System.Windows.Forms.AnchorStyles.Bottom | System.Windows.Forms.AnchorStyles.Right);
			this.btnSend.Location = new System.Drawing.Point(385, 302);
			this.btnSend.Name = "btnSend";
			this.btnSend.Size = new System.Drawing.Size(75, 37);
			this.btnSend.TabIndex = 1;
			this.btnSend.Text = "Send";
			this.btnSend.Click += new System.EventHandler(this.btnSend_Click);
			// 
			// wfMsg
			// 
			this.AcceptButton = this.btnSend;
			this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
			this.ClientSize = new System.Drawing.Size(461, 341);
			this.Controls.AddRange(new System.Windows.Forms.Control[] {
																		  this.btnSend,
																		  this.txtChat,
																		  this.txtMsg});
			this.Icon = ((System.Drawing.Icon)(resources.GetObject("$this.Icon")));
			this.Name = "wfMsg";
			this.Text = "MessageTemplate";
			this.Enter += new System.EventHandler(this.wfMsg_Enter);
			this.ResumeLayout(false);

		}
		#endregion

		public void ShowWindow()
		{
			// Show the window, and add the text
			this.Show();
			this.Select();
			this.BringToFront();
		}
		private void wfMsg_Enter(object sender, System.EventArgs e)
		{
			// We have the focus, make the chat textbox the default option
			this.txtChat.Select();
		}

		public void AddChatMessage(string msg, bool meTalking, bool noTalking)
		{
			if (!noTalking)
			{
				if (meTalking)
					msg = "<" + parent.gUsername + "> " + msg;
				else
					msg = "<" + msUser + "> " + msg;
			}

			// Now limit the text to 32k
			if (txtMsg.Text.Length > (txtMsg.MaxLength * 0.95))
				txtMsg.Text = txtMsg.Text.Remove(0, (int)(txtMsg.MaxLength / 2));

			// Update the message window
			txtMsg.Text += msg + "\r\n";
			txtMsg.SelectionStart = txtMsg.Text.Length;
			txtMsg.ScrollToCaret();

		}
		private void btnSend_Click(object sender, System.EventArgs e)
		{
			// We're talking
			if ((txtChat.Text != null) && (txtChat.Text != ""))
			{
				if (parent.gConnected)
				{
					parent.SendChatMessage(msUser, txtChat.Text);
					object[] unavailableMsg = {txtChat.Text, true, false };

					this.BeginInvoke(new AddChatMsgCallback(this.AddChatMessage), unavailableMsg);
					txtChat.Text = null;
				}
				else
				{
					object[] unavailableMsg = {"**** - You are not connected to a server, you cannot send messages.", true, true };
					this.BeginInvoke(new AddChatMsgCallback(this.AddChatMessage), unavailableMsg);
					txtChat.Text = null;
				}
			}
		}

		private void txtChat_TextChanged(object sender, System.EventArgs e)
		{
			// Only allow them to send if there's data to be sent.
			btnSend.Enabled = ((txtChat.Text != null) && (txtChat.Text != ""));
		}
	}
}
